### ISSUE TITLE
XFetchName doesn't always return a name.

### ISSUE THREAD

**Issue by prohunterdogkeeper** [herbstluftwm/herbstluftwm#64]
Even though a window might have a WM_NAME property, XFetchName doesn't always return a name - for some reason. Googling around a bit I couldn't figure out why though. I noticed that herbstluftwm didn't name Steam's windows. Steam windows don't have a _NET_WM_NAME property, so herbstluftwm falls back to calling XFetchName to get the WM_NAME property.
Running herbstluftwm in verbose, it prints that "no title for window xxx found, using ''" when a Steam window is focused, where xxx is the ID of a Steam window. This means that XFetchName returned 0, even though using xprop shows the window to indeed have a WM_NAME property. 

I did a test of the XGetWMName function, using code from this stack overflow thread [1], which returned the correct name, where XFetchName returned no name.

I propose to change XFetchName to XGetWMName.

[1] https://stackoverflow.com/questions/8825661/xfetchname-always-returns-0

**Comment by ypnos** [comment]
To me this looks like a very reasonable suggestion.
It needs to be noted that XGetWMName explicitely deals with encoding and provides a way to query for the encoding of the returned value, while XFetchName's result is "implementation dependent". To quote:

> f the data returned by the server is in the Latin Portable Character Encoding, then the returned string is in the Host Portable Character Encoding. Otherwise, the result is implementation dependent.

"Implementation dependent" might very well mean returning zero.

Could you maybe prepare a pull request for the change?

**Comment by prohunterdogkeeper** [comment]
I'm not much of a programmer, and so I haven't set up a git workflow, neither do I feel very confident in my programming abilities, but wouldn't something as simple as this work? 
```
XTextProperty xtp_new_name;\n
/* fall back to WM_Name */
if (0 != XGetWMName(g_display, client->window, &xtp_new_name)) {
    new_name = g_string_new(xtp_new_name.value);
```
This replacing lines 871 through 875 in src/clientlist.cpp

I'm not familiar with xlib, and so I assume you don't need to XFree() the XTextProperty object, since it doesn't mention it in the manpage for XGetWMName, while it does mention it in the description of XFetchName?

**Comment by ypnos** [comment]
Yes this looks o.k. however based on the documentation of XTextProperty I think one needs to not directly access `value` but use conversion methods instead. Maybe thorsten knows better how to do this?

https://tronche.com/gui/x/xlib/ICC/client-to-window-manager/converting-string-lists.html

**Comment by prohunterdogkeeper** [comment]
I did some testing, and it would _seem_ that using the code I posted would work without regressions, I haven't tried compiling herbstluft with it though.

I don't think I understand character encoding though. This method works with 'regular' European unicode characters (tested with æøå), but not with Japanese characters. Current way doesn't work with Japanese characters either, might be worth fixing that as well now though.

I've been messing a bit with XwcTextPropertyToTextList() from your link, but I can't get it to not segfault.

**Comment by ypnos** [comment]
After quite a while of digging, I found this code:
https://code.google.com/p/chromium/codesearch#chromium/src/third_party/webrtc/base/x11windowpicker.cc&q=xtextproperty&sq=package:chromium&type=cs&l=639

I didn't know of `Xutf8TextPropertyToTextList` before, but I found that it already used by hlwm IPC code. So it is safe to use. Would you mind testing it?

**Comment by prohunterdogkeeper** [comment]
I think I fixed it :)

**Comment by ypnos** [comment]
Thank you for the good work!
Fixed in e8a9ec45e769f49cc4d846f98ac7a7f35258bc30.
